src/OracleService/Protocols/OracleHttpsProtocol.cs

```diff
@@ -48,17 +48,24 @@ public async Task<(OracleResponseCode, string)> ProcessAsync(Uri uri, Cancellati
         {
             Utility.Log(nameof(OracleHttpsProtocol), LogLevel.Debug, $"Request: {uri.AbsoluteUri}");
 
-            if (!Settings.Default.AllowPrivateHost)
-            {
-                IPHostEntry entry = await Dns.GetHostEntryAsync(uri.Host);
-                if (entry.IsInternal())
-                    return (OracleResponseCode.Forbidden, null);
-            }
-
             HttpResponseMessage message;
             try
             {
-                message = await client.GetAsync(uri, HttpCompletionOption.ResponseContentRead, cancellation);
+                do
+                {
+                    if (!Settings.Default.AllowPrivateHost)
+                    {
+                        IPHostEntry entry = await Dns.GetHostEntryAsync(uri.Host, cancellation);
+                        if (entry.IsInternal())
+                            return (OracleResponseCode.Forbidden, null);
+                    }
+                    message = await client.GetAsync(uri, HttpCompletionOption.ResponseContentRead, cancellation);
+                    if (message.Headers.Location is not null)
+                    {
+                        uri = message.Headers.Location;
+                        message = null;
+                    }
+                } while (message == null);
```

Comment:
So it's an infinite loop, right?

Reply:
There are a timeout

Reply:
Every `client.GetAsync` will have its own timeout. But there is no timeout for the whole loop.

What if some url redirects to itself immediately?

Reply:
Reviewing the code it's right, I thought that the cancelation token was used as timeout. My bad, I will fix it.